### ISSUE TITLE
Visualization: Custom Charts

### ISSUE THREAD

**Issue by rohit-guitar** [metabase/metabase#2318]
Can i add my custom chart type in metabase?

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

**Comment by tlrobinson** [comment]
Two things:

First, I'm curious what would your ideal way of adding custom charts be? Do you want:

1. to build and upload fully custom visualizations using arbitrary JavaScript libraries you provide (D3, React, etc)
2. to use a more limited declarative API such as vega or vega-lite which we could provide an in-app editor for
3. other?

Second, we don't currently have a supported way to add custom charts. That said, if you're willing to recompile, and ok with the possibility the API might break with future updates, here are the basic steps:

1. Create a new React component, here: https://github.com/metabase/metabase/tree/master/frontend/src/visualizations
2. Import and register it here: https://github.com/metabase/metabase/blob/master/frontend/src/visualizations/index.js
3. Rebuild Metabase, instructions here: https://github.com/metabase/metabase/blob/master/docs/developers-guide.md

Basically the React component takes a prop `data` that has a `cols` array containing metadata about the returned columns, and a `rows` array containing each rows' values as another array, i.e.

```
data: {
  cols: [
    { name: "column1", display_name: ..., base_type: ... },
    { name: "column2", display_name: ..., base_type: ... }
  ],
  rows: [
    ["a", 1],
    ["b", 2],
    ["c", 3],
    ...
  ]
}
```

Also note we need to do #2205 before you'll be able to save cards with new chart types, but that's an easy fix. Let me know if you need help with that.

**Comment by camsaul** [comment]
@tlrobinson it's inaccurate to say that #2205 needs to be done before someone can add a custom chart type. Custom charts will require changes to the source code regardless of whether or not the backend data validation code is removed; it's trivial to add new types to the list of valid visualization types.

**Comment by tlrobinson** [comment]
@camsaul Sure, but we should do #2205 so in the future people adding custom charts don't need to worry about it.

**Comment by siscia** [comment]
@tlrobinson I am not OP but I second that custom charts are very interesting.

In my opinion I would like to pass a js file that define a single function that takes as input the result of the query.

Most likely, will be necessary to declare what shape of data are accepted from the visualization.

**Comment by joshreuben456** [comment]
👍

**Comment by robs15** [comment]
This is still open. I'm interested in using this framework & would like to know if I can go ahead & build a custom chart as per the instructions by @tlrobinson ?

Thanks in advance.

**Comment by ryanmccoy** [comment]
@tlrobinson in response to your first post and question to the OP, and as an extremely interested party in this feature request:

I would love the ability to have both an in-app editor for declarative tools like Vega AND the ability to upload my own custom visualizations with the likes of D3, but if I had to choose one I would choose something like Vega for sure.

**Comment by timuric** [comment]
Having Vega charts would be a great start!

**Comment by rayterrill** [comment]
A potential workaround for us would be some ability to add custom html/javascript as a "widget" on the dashboard to implement a chart or visualization not currently supported by Metabase. Something we'd love - some way to do a very simple, dynamic "thumbs up" or "thumbs down" visualization based on the results returned (i.e. thumbs down if current operations throughput is below 5 parts/day, with both the current value and the target being dynamic/returned via the result set if possible).

**Comment by trcull** [comment]
Hey guys, it's been a while since the last comment on this and wondering what the thinking is?  I might even be willing to do the coding myself since I need this so badly, but wouldn't want to do that if it's something the maintainers don't actually like or have an appetite for.

I recognize that many other features have more votes than this one, but at the same time this feature would be a great "get out of jail free" card that frees people up to solve their own problems without waiting for releases.  For example, this feature (https://github.com/metabase/metabase/issues/5474) has a lot of upvotes, but if there were a custom chart type then its OP could actually just do it themselves with d3. 

If I were the one doing it, the basic design would be:
1) custom chart config in Metabase takes a) a textbox with any js libraries and css files it should import at most once per dashboard/question page, b) a textbox full of arbitrary js it should place in a script tag where the chart goes, c) the name of a javascript class and callback method it should expect to find in the window context, d) maybe a css class name or something to apply to the containing div, oh and e) the query, of course
2) at chart load time, the Metabase engine executes the query as normal and calls the configured js class/method and hands it the data structure @tlrobinson mentioned above:

data: {
  cols: [
    { name: "column1", display_name: ..., base_type: ... },
    { name: "column2", display_name: ..., base_type: ... }
  ],
  rows: [
    ["a", 1],
    ["b", 2],
    ["c", 3],
    ...
  ]
}

And that's it.  Everything else would be the chart implementor's problem.

**Comment by dwelch2344** [comment]
@trcull That's one way, but personally it feels pretty nasty/hacky. Providing Jars and compile time solutions also are not ideal, imo. Version control, debugging, etc all become pretty painful when there's a lot of moving pieces to get setup. 

Perhaps some way to simply specify client-side scripts (in the config, manifest, etc) via a browser URL + some sort of mechanism for the scripts to load themselves per their needs? Something along the lines of having a client side page load lifecycle with "cutpoints" if you will. 

Rough idea of flow: 
1) Page render begins, with the javascript block of metabase loaded and a global object in the window created. Say `Metabase`, `mbplugins` something (we'll find a safe name) 
2) During page render, after said global object is available in context, plugin files are loaded via `<script src="...">` tags, per the ordering and fully qualified URL configured in admin 
3) Plugins can do what they need, but are aware that there *should* be a global object `Metabase` to act as a registry, and they can do something like the following: 
```js
Metabase.init.register(function metabasePluginHook(metabaseContextOrSomething){
   metabaseContextOrSomething.addChart('someName', reactFunctionRef, {etc}) 
})
```
4) Once the "right" time has come (i.e. Metabase client-side is loaded to whatever level is needed) the plugins are iterated and initialized, passing in whatever state they need to make Metabase aware and do smart-and-useful-things.



This clearly needs thought from a security model standpoint, but the pattern is also well known and utilized. Microsoft uses this model now for all their office plugins (both Desktop and Office365 on web) and it's actually super nice to work with

**Comment by CrazyZhangsk** [comment]
@tlrobinson Has a JavaScript library or other components been provided to add charts?

**Comment by howardmalec** [comment]
There's ~170 open issues related to visualizations. Metabase's viz capabilities are limited to a small set of chart types (lines, tables, bars, maps, big numbers), especially when compared to alternative platforms like Mode and Apache Superset. 

Mirroring Apache Superset and Mode, Metabase should consider implementing a visualization plug-in architecture that makes it easy for users to add custom visualizations to Metabase. Examples include:

- [Superset](https://superset.apache.org/docs/installation/building-custom-viz-plugins/)
- [Mode](https://mode.com/blog/custom-charts-gallery/)

Right now every new visualization requires an issue to be filed and prioritized as a feature request. Metabase is missing many must-have viz types like [box-plots](https://github.com/metabase/metabase/issues/10783) and [sparklines](https://github.com/metabase/metabase/issues/10338). 

Enabling a self-service architecture to let users install custom visualizations from libraries like d3, kepler, etc. would eliminate a significant # of requests for custom viz that are currently requested as one-off implementations. 

Limited visualization capabilities are a major reason to consider alternatives over Metabase. 

@flamber - any chance this will be prioritized given the multiple ongoing threads requesting this feature, and the large # of tickets asking for custom viz?

**Comment by timuric** [comment]
metabase + https://observablehq.com/ would be ground breaking!

**Comment by howardmalec** [comment]
@flamber - Wanted to get your thoughts.

According to the metabase roadmap (<= .39) the team is active working on multiple chart viz requests:

- [Waterfall Charts
](https://github.com/metabase/metabase/issues/11904)
- [Chart Grid Heatmap](https://github.com/metabase/metabase/issues/5474)
- [Chart Cohort](https://github.com/metabase/metabase/issues/3188)

All of these requests and other outstanding asks like box plots, sparklines, etc could be addressed by end-users w/o customer integrations if metabase focused on building an extensible visualization framework like Mode or Superset.

It could be worth considering this request as higher priority vs. the other current one-off viz integrations currently on the roadmap.

**Comment by holgerbrandl** [comment]
Another great source for inspiration could be https://grafana.com/docs/grafana/latest/developers/plugins/ where they expose a panel extension API. 

Having some nice market-place interface such as https://grafana.com/grafana/plugins/?type=panel on top of it would be awesome. However, the first step should imho be an extension API for custom panels.

**Comment by ayazwani** [comment]
I'm trying to add custom charts into the metabase. Has anyone done any work on the same. Please do share some experience and resources.I want to start with for doing it. Or if already done share the GitHub repo or any article or blog for the same.
Charts i want to add:
1: Box plots [https://www.d3-graph-gallery.com/graph/boxplot_several_groups.html](url)
2: Sankey charts [https://www.d3-graph-gallery.com/graph/sankey_basic.html](url)

**Comment by Aryan779** [comment]
I am not an expert at javascript frameworks but can someone help me to understand the code, as I want to add calendar heatmap in visualization.

**Comment by trcull** [comment]
I'm currently at a totally different company than I was in 2019, when I commented above ^^ and we also have a need for custom charts and could use this plugin architecture.  I'd forgotten I'd even commented on this issue five years ago and just came upon it again while trying to figure out how to do it.  Please, please add this!  Or let someone else do it!  Lacking this visual plugin ability is a major barrier to adoption (it's stopped me from recommending Metabase, which I otherwise love, at several different organizations now)

**Comment by RoyWiggins** [comment]
imho the minimum version of this feature would be something like Redash, which just gives you a textbox that you can just paste Javascript into to define a Plotly plot. We've found this to be flexible enough to produce nice, well-integrated plots that blend in with the builtin ones, and it would be very useful if Metabase could also do something like this:

![Image](https://github.com/user-attachments/assets/925c6c9f-83ce-4df8-a4b6-2ea122d6f8d1)

It's not a full plugin architecture, or even a UI to let you define your own plot types- if you want two visualizations using the same custom code you have to copy and paste. It's very, very minimal, but it's much better than nothing.

**Comment by chippyash** [comment]
Sad that this hasn't gained any traction in 9+ yrs. !! The ability to add custom chart types would make Metabase all that more attractive. For instance, I want to be able to create directed graph chart. Metabase can't, Superset can.  We use graphDb products like Neo4J and Memgraph. Whilst I can produce DAGs with the graph tools, I can't surface them to potential consumers via Metabase, even in a simplified format.

**Comment by dwelch2344** [comment]
Agreed w previous. If nothing else, there should be a pluggable way to provide your own (visualizations, or even just UI components) which afaict there isn't. 


I'd totally be willing to volunteer if I could get guidance on what would be acceptable.

**Comment by dragonsahead** [comment]
@chippyash @dwelch2344 feel free to contribute https://www.metabase.com/docs/latest/developers-guide/start, we're an open source product

**Comment by psarin** [comment]
@paoliniluis @tlrobinson Any updates on this one? I looked at instructions above but it seems as if there have been significant changes since then and the instructions don't apply anymore. Like @dwelch2344 , would be great to get some guidance on what specific things we can do to contribute positively to this requests. Thanks.

**Comment by psarin** [comment]
@dragonsahead @tlrobinson Any thoughts on how we can contribute towards this feature (see https://github.com/metabase/metabase/issues/2318#issuecomment-3419137515)? Thanks.

**Comment by kamilmielnik** [comment]
We plan to release custom visualizations in Metabase 62.
The feature will be available in Pro and Enterprise plans.

We've built 2 reference custom visualization implementations:
- https://github.com/metabase/custom-viz-thumbs (thanks @rayterrill for [the comment](https://github.com/metabase/metabase/issues/2318#issuecomment-436445697))
- https://github.com/metabase/custom-viz-calendar-heatmap
